action/protocol/staking/protocol.go

```diff
@@ -468,6 +468,52 @@ func (p *Protocol) Validate(ctx context.Context, act action.Action, sr protocol.
 	return nil
 }
 
+func (p *Protocol) isActiveCandidate(ctx context.Context, csr CandidateStateReader, cand *Candidate) (bool, error) {
+	// at least min self stake
+	if cand.SelfStake.Cmp(p.config.RegistrationConsts.MinSelfStake) < 0 {
+		return false, nil
+	}
+
+	// endorsement must not exipred if the self-stake bucket is an endorse bucket
+	featureCtx := protocol.MustGetFeatureCtx(ctx)
+	if featureCtx.DisableDelegateEndorsement {
+		return true, nil
+	}
+	srHeight, err := csr.SR().Height()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get StateReader height")
+	}
+	vb, err := csr.getBucket(cand.SelfStakeBucketIdx)
+	if err != nil {
+		if errors.Is(err, state.ErrStateNotExist) {
+			return false, nil
```

Comment:
SelfStakeBucketIndex shouldn't be an invalid one, so we may always return `false, err`